File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isAlpha, isSpace, isUpper)

responseFor :: String -> String
responseFor x 
    | all isSpace x = "Fine. Be that way!"
    | charsAreUppercase && atLeastOneCharIsUppercase && last x == '?' = "Calm down, I know what I'm doing!"
    | charsAreUppercase && atLeastOneCharIsUppercase = "Whoa, chill out!"
    | last (trim x) == '?' = "Sure."
    | otherwise = "Whatever."
    where charsAreUppercase = all isUpper (filter isAlpha x)
          atLeastOneCharIsUppercase = any isUpper x

trim :: String -> String
trim = f . f
   where f = reverse . dropWhile isSpace

File: cpp/bob/bob.cpp
#include "bob.h"
#include <string>

namespace bob {
  std::string hey(std::string msg) {
    msg.erase(msg.find_last_not_of(" \n\r\t")+1);
    msg.erase(0, msg.find_first_not_of(" \n\r\t"));

    if(msg.length() == 0) {
      return "Fine. Be that way!";
    }

    if (has_alpha(msg) && to_upper(msg) == msg) {
      if (msg[msg.length() - 1] == '?') {
        return "Calm down, I know what I'm doing!";
      }
      return "Whoa, chill out!";
    }

    if (msg[msg.length() - 1] == '?') {
      return "Sure.";
    }

    return "Whatever.";
  }

  std::string to_upper(std::string str) {
    std::string result{""};
    for (int x=0; x < (int) str.length(); x++) {
      result += std::toupper(str[x]);
    }

    return result;
  }

  bool has_alpha(std::string str) {
    for (int x=0; x < (int) str.length(); x++) {
      if (isalpha(str[x])) { return true; }
    }
    return false;
  }
}
